### DISCUSSION TITLE
Porting raylib to `wgpu`

### DISCUSSION THREAD

**Discussion by AregevDev** [raysan5/raylib#4505] [Features]
# Context
`WebGPU` is a new graphics API that aims to be a modern graphics API on the web. `wgpu` is one of the available implementations of `WebGPU`, written in Rust, and used internally in Firefox. `wgpu` currently translates WebGPU calls to `DirectX 11`, `DirectX 12`, `Vulkan`,  `Metal`, and `OpenGL` / `GLES` / `WebGL`. `wgpu-native` is the C binding library for `wgpu`, usable from C / C++.

As an exercise, I attempt to port raylib to use `wgpu`. This proves to be a challenge since raylib is deeply dependent on `rlgl` and `OpenGL` itself. Virtually, every raylib core module uses `rlgl`. My goal is to split the core modules from the graphics backend and write a new `wgpu` graphics backend. This can help future-proofing raylib, while still supporting it on older systems. **I do not intend to replace `rlgl` completely.**

# How?
To write a new graphics backend, I probably need to split/abstract the graphics initialization.
Some ideas I had in mind:
- Add extra platform source files for the wgpu implementation (i.e `rcore_desktop_glfw_wgpu`) This approach is similar to what `ImGui` does, by letting the user include the right combination of window platform and rendering backend. This approach might not fully work since raylib is tightly integrated with `OpenGL`.
- Abstract the `rlgl` backend. Make all `raylib.h`'s functions not depend on `rlgl` directly, but rather select the correct graphics backend based on the build configuration.
- Implement the `rlgl` functions in `wgpu`. To be honest, I am not sure about that solution. Mapping old-style `OpenGL` calls to modern graphics APIs can lead to suboptimal usage/performance.

This proves to be an extremely critical decision from a developer experience and simplicity standpoint.

# I need help
I opened this discussion to gain some feedback and start a discussion.
For now, I modified the CMake scrips to download `wgpu-native` and build it from source. You can view the changes here: https://github.com/AregevDev/raylib/tree/wgpu.

**Comment by JeffM2501** [comment]
Would this add rust as a dependency to raylib?

**Reply by AregevDev** [reply]
Not necessarily, I intend to offer prebuilt versions from their release page: https://github.com/gfx-rs/wgpu-native/releases
For building from source, yeah you would need Rust and `libclang`.

I understand the concern, the default behavior can use the prebuilt binaries from the release page instead of forcing everyone to install Rust.

You can always disable the backend if you are not interested in it. (For old console builds for example)

**Reply by manuel5975p** [reply]
Or you could make it dawn (C++) based, which will be also what emscripten uses.

**Reply by wwderw** [reply]
> Or you could make it dawn (C++) based, which will be also what emscripten uses.

Chrome uses it as well (big surprise there).  I think that's what Mach (for Zig) uses as well, unless their own inhouse gets further along.

Given Raylib uses Angle for Vulkan, maybe the way to go would be with Dawn, I dunno, I just know enough to cause damage.

**Reply by JeffM2501** [reply]
> Not necessarily, I intend to offer prebuilt versions from their release page: https://github.com/gfx-rs/wgpu-native/releases For building from source, yeah you would need Rust and `libclang`.
> 
> I understand the concern, the default behavior can use the prebuilt binaries from the release page instead of forcing everyone to install Rust.
> 
> You can always disable the backend if you are not interested in it. (For old console builds for example)

That's a big downside if rust is a dependency just to build raylib. Would it be 100% optiona? There are many many users who build from source and this adds a lot.

Are there web GPU versions that do not depend on rust?

**Reply by manuel5975p** [reply]
@JeffM2501 as stated, there is dawn, which is only C/C++ based.

**Reply by AregevDev** [reply]
Building dawn from source requires many "google"-like dependencies: `depot_tools`, `GN` etc. I find the C bindings of `wgpu` more pleasant to work with. As for web, `wgpu` supports both `WebGL` and `WebGPU` on wasm.

**Reply by manuel5975p** [reply]
It doesn't require depot_tools, only the python script, which I'm sure I could fix with some CMake / git trickery. Also the GN builds are no longer maintained afaik, everything is buildable with CMake.

**Reply by gen2brain** [reply]
Sorry to intrude on the discussion, but prebuilt binaries, for what, only Windows, macOS, and Linux, and only amd64? That looks like trouble. The main advantage is that you can use whatever C compiler/toolchain for whatever platform, and there are ongoing discussions and work for software-only support. Something like this ruins everything.

**Reply by wwderw** [reply]
> Sorry to intrude on the discussion, but prebuilt binaries, for what, only Windows, macOS, and Linux, and only amd64? That looks like trouble. The main advantage is that you can use whatever C compiler/toolchain for whatever platform, and there are ongoing discussions and work for software-only support. Something like this ruins everything.

As someone that is looking forward to software-only support, I do have that concern as well.

**Reply by AregevDev** [reply]
Really? That's cool.
So the idea is to provide an external library for custom rendering backends. Then, the only thing to worry about in the main repository is the core-graphics split.

**Reply by JeffM2501** [reply]
In general we don't want any new dependencies unless they are absolutely necessary. Raylib is built with MANY different build systems by a wide range of people. We can not assume any specific libraries or tools or prebuild binaries. That even includes tooling such as python.

**Reply by AregevDev** [reply]
Yeah, fair.

**Comment by ghost** [comment]
Reasons against changing raylib to accommodate this proposal:

1. WebGPU is experimental ([ref](https://developer.mozilla.org/en-US/docs/Web/API/WebGPU_API)), has limited availability ([ref](https://developer.mozilla.org/en-US/docs/Web/API/WebGPU_API#browser_compatibility)) and is still a working draft ([ref](https://www.w3.org/TR/webgpu/)).
2. It would require a substancial rework of nearly everything in raylib (right after it finally reached a point of good stability).
3. Raylib would add a dependency on Rust (which is very far away from C99).

**Reply by manuel5975p** [reply]
I agree mostly but again, there is a pure C/C++ webgpu implementation called [Dawn](https://dawn.googlesource.com/)

**Reply by AregevDev** [reply]
@jeanck All fair points, I should not be asking for wgpu in the core library. Maybe reorganize the source files to make writing extra graphics backends more approachable.

(My inner Tsoding is yelling at me rn 😄)

**Reply by adex345** [reply]
Why would browser support be a problem in raylib? Does it support browsers already?

**Comment by OetkenPurveyorOfCode** [comment]
I once thought of implementing a raylib platform backend (similar to GLFW, SDL2, RGFW) with sokol. 
Andre Weissflog's sokol_gfx provides abstractions for different graphics backends (including WebGPU), namely DirectX 11, OpenGL, OpenGL ES, Metal, WebGL, WebGPU. 
But is it worth it? I don't know.
It would have many of the same problems that supporting WebGPU would have. The biggest problem would be the shaders. A raylib application would have to supple the shader in the shading language of the target platform. (Sokol has a cross-compiler from Vulkan GLSL to all its backends, but that does not cover the cases when one wants to generate shaders at runtime).
